source/mir/base64.d

```diff
@@ -0,0 +1,317 @@
+
+        // We do the inverse of how we encoded it...
+        uint transformed_group = (a << 26) | (b << 20) | (c << 14) | (d << 8);
+
+        const(ubyte) t_a = (transformed_group >> 24) & 0xff;
+        const(ubyte) t_b = (transformed_group >> 16) & 0xff;
+        const(ubyte) t_c = (transformed_group >> 8) & 0xff;
+        const(ubyte) t_d = (transformed_group) & 0xff;
+
+        // We should *always* have enough for at least
+        // one, but we don't need to have enough for the rest..
+        appender.put(t_a);
+
+        // Only emit transformed groups if we have enough data for them.
+        if (t_b == 0 && t_c == 0 && t_d == 0)
+        {
+            return;
+        }
+        else if (t_c == 0 && t_d == 0)
+        {
+            appender.put(t_b);
+        } 
+        else if (t_d == 0)
+        {
+            appender.put(t_b);
+            appender.put(t_c);
+        }
+        else
+        {
+            appender.put(t_b);
+            appender.put(t_c);
+            appender.put(t_d);
+        }
+    }
+}
+
+/// Test decoding of data which has a length which can be
+/// cleanly decoded.
+unittest
+{
+    {
+        ubyte[] data = cast(ubyte[])"QUJD";
+        assert(data.decodeBase64 == "ABC");
+    }
+
+    {
+        ubyte[] data = cast(ubyte[])"QQ==";
+        assert(data.decodeBase64 == "A");
+    }
+
+    {
+        ubyte[] data = cast(ubyte[])"YSBiIGMgZCBlIGYgZyBoIGkgaiBrIGwgbSBuIG8gcCBxIHIgcyB0IHUgdiB3IHggeSB6";
+        assert(data.decodeBase64 == "a b c d e f g h i j k l m n o p q r s t u v w x y z");
+    }
+
+    {
+        ubyte[] data = cast(ubyte[])"LCAuIDsgLyBbICcgXSBcID0gLSAwIDkgOCA3IDYgNSA0IDMgMiAxIGAgfiAhIEAgIyAkICUgXiAmICogKCApIF8gKyB8IDogPCA+ID8=";
+        assert(data.decodeBase64 == ", . ; / [ ' ] \\ = - 0 9 8 7 6 5 4 3 2 1 ` ~ ! @ # $ % ^ & * ( ) _ + | : < > ?");
+    }
+}
+
+/++
+Encode a ubyte array as Base64, returning the encoded value.
++/
+ubyte[] encodeBase64(scope ubyte[] buf) @safe pure
+{
+    import mir.appender : scopedBuffer;
+    auto app = scopedBuffer!ubyte;
+    encodeBase64(buf, app);
+    return app.data.dup;
+}
+
+/++
+Encode a ubyte array as Base64, placing the result onto an Appender.
++/
+void encodeBase64(Appender)(scope ubyte[] input, ref Appender appender) @safe @nogc pure
+{
+    // Slice our input array so that n % 3 == 0 (we have a multiple of 3) 
+    // If we have less then 3, then this is effectively a no-op (will result in a 0-length slice)
+    ubyte[] window = input[0 .. input.length - (input.length % 3)];
+    foreach(group; window.bytegroup!(3, uint).map!bswap) {
+        const(ubyte) a = (group >> 26) & 0x3f;
+        const(ubyte) b = (group >> 20) & 0x3f;
+        const(ubyte) c = (group >> 14) & 0x3f;
+        const(ubyte) d = (group >> 8) & 0x3f;
+
+        appender.put(a.lookup_encoding);
+        appender.put(b.lookup_encoding);
+        appender.put(c.lookup_encoding);
+        appender.put(d.lookup_encoding);
```

Comment:
minimize amount of `appender.put` calls. It should be just `appender.put(endodedByteGroup[])`, where `ubyte[4] endodedByteGroup`